flink-table-store-core/src/main/java/org/apache/flink/table/store/file/schema/SchemaManager.java

```diff
@@ -0,0 +1,144 @@
+
+    /** Default no lock. */
+    private Lock lock = Callable::call;
+
+    public SchemaManager(Path tableRoot) {
+        this.tableRoot = tableRoot;
+    }
+
+    public SchemaManager withLock(Lock lock) {
+        this.lock = lock;
+        return this;
+    }
+
+    /** @return latest schema. */
+    public Optional<Schema> latest() {
+        try {
+            return listVersionedFiles(schemaDirectory(), SCHEMA_PREFIX)
+                    .reduce(Math::max)
+                    .map(this::schema);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /** Create a new schema from {@link UpdateSchema}. */
+    public Schema commitNewVersion(UpdateSchema updateSchema) throws Exception {
+        RowType rowType = updateSchema.rowType();
+        List<String> partitionKeys = updateSchema.partitionKeys();
+        List<String> primaryKeys = updateSchema.primaryKeys();
+        Map<String, String> options = updateSchema.options();
+
+        while (true) {
+            long id;
+            int highestFieldId;
+            List<DataField> fields;
+            Optional<Schema> latest = latest();
+            if (latest.isPresent()) {
+                Schema oldSchema = latest.get();
+                Preconditions.checkArgument(
+                        oldSchema.primaryKeys().equals(primaryKeys),
+                        "Primary key modification is not supported, "
+                                + "old primaryKeys is %s, new primaryKeys is %s",
+                        oldSchema.primaryKeys(),
+                        primaryKeys);
+
+                if (!updateSchema
+                                .rowType()
+                                .getFields()
+                                .equals(oldSchema.logicalRowType().getFields())
+                        || !updateSchema.partitionKeys().equals(oldSchema.partitionKeys())) {
+                    throw new UnsupportedOperationException(
+                            "TODO: support update field types and partition keys. ");
+                }
+
+                fields = oldSchema.fields();
+                id = oldSchema.id() + 1;
+                highestFieldId = oldSchema.highestFieldId();
+            } else {
+                fields = Schema.newFields(rowType);
+                highestFieldId = Schema.currentHighestFieldId(fields);
+                id = 0;
+            }
+
+            Schema schema =
+                    new Schema(id, fields, highestFieldId, partitionKeys, primaryKeys, options);
+
+            Path temp = toTmpSchemaPath(id);
+            Path finalFile = toSchemaPath(id);
+            FileUtils.writeFileUtf8(temp, schema.toString());
+
+            Boolean success = lock.runWithLock(() -> temp.getFileSystem().rename(temp, finalFile));
+            if (success) {
+                return schema;
+            } else {
+                // retry
+                FileUtils.deleteOrWarn(temp);
```

Comment:
If this method fails with exception `temp` file will not be cleaned.

Reply:
Yes, but we dont have other solutions.

Reply:
You have. Wrap the code with `try... finally...`.